File: internal/log_analysis/rules_engine/src/sqs.py
# Panther is a scalable, powerful, cloud-native SIEM written in Golang/React.
# Copyright (C) 2020 Panther Labs Inc
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import json
import os
from datetime import datetime
from typing import List, Dict

import boto3

# Max number of SQS messages inside an SQS batch
_MAX_MESSAGES = 10
# Max size of an SQS batch request
_MAX_MESSAGE_SIZE = 256 * 1000

sqs_resource = boto3.resource('sqs')
queue = sqs_resource.get_queue_by_name(QueueName=os.environ['ALERTS_QUEUE'])


def send_to_sqs(matches: List) -> None:
    """Send a tuple of (rule_id, event) to SQS."""
    messages = [match_to_sqs_entry_message(i) for i in matches]

    current_entries: List[Dict[str, str]] = []
    current_byte_size = 0

    for i in range(len(messages)):
        entry = {'Id': str(i), 'MessageBody': messages[i]}
        projected_size = current_byte_size + len(messages[i])
        projected_num_entries = len(current_entries) + 1
        if projected_num_entries > _MAX_MESSAGES or projected_size > _MAX_MESSAGE_SIZE:
            queue.send_messages(Entries=current_entries)
            current_entries = [entry]
            current_byte_size = len(messages[i])
        else:
            current_entries.append(entry)
            current_byte_size += len(messages[i])

    if len(current_entries) > 0:
        queue.send_messages(Entries=current_entries)

    return


def match_to_sqs_entry_message(match: (str, str)) -> str:
    notification = {'ruleId': match[0], 'event': match[1], 'timestamp': datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')}
    return json.dumps(notification)

File: internal/core/alert_delivery/outputs/sqs.go
package outputs

/**
 * Panther is a scalable, powerful, cloud-native SIEM written in Golang/React.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import (
	"strings"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/sqs"
	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
	jsoniter "github.com/json-iterator/go"
	"go.uber.org/zap"

	outputmodels "github.com/panther-labs/panther/api/lambda/outputs/models"
	alertmodels "github.com/panther-labs/panther/internal/core/alert_delivery/models"
)

// Sqs sends an alert to an SQS Queue.
// nolint: dupl
func (client *OutputClient) Sqs(alert *alertmodels.Alert, config *outputmodels.SqsConfig) *AlertDeliveryError {
	outputMessage := &sqsOutputMessage{
		ID:          alert.PolicyID,
		Name:        alert.PolicyName,
		VersionID:   alert.PolicyVersionID,
		Description: alert.PolicyDescription,
		Runbook:     alert.Runbook,
		Severity:    alert.Severity,
		Tags:        alert.Tags,
	}

	serializedMessage, err := jsoniter.MarshalToString(outputMessage)
	if err != nil {
		zap.L().Error("Failed to serialize message", zap.Error(err))
		return &AlertDeliveryError{Message: "Failed to serialize message"}
	}

	sqsSendMessageInput := &sqs.SendMessageInput{
		QueueUrl:    config.QueueURL,
		MessageBody: aws.String(serializedMessage),
	}

	sqsClient, err := client.getSqsClient(*config.QueueURL)
	if err != nil {
		return &AlertDeliveryError{Message: "Failed to create Sqs client for queue", Permanent: true}
	}

	_, err = sqsClient.SendMessage(sqsSendMessageInput)
	if err != nil {
		zap.L().Error("Failed to send message to SQS queue", zap.Error(err))
		return &AlertDeliveryError{Message: "Failed to send message to SQS queue"}
	}
	return nil
}

//sqsOutputMessage contains the fields that will be included in the SQS message
type sqsOutputMessage struct {
	ID          *string   `json:"id"`
	Name        *string   `json:"name,omitempty"`
	VersionID   *string   `json:"versionId,omitempty"`
	Description *string   `json:"description,omitempty"`
	Runbook     *string   `json:"runbook,omitempty"`
	Severity    *string   `json:"severity"`
	Tags        []*string `json:"tags,omitempty"`
}

func (client *OutputClient) getSqsClient(queueURL string) (sqsiface.SQSAPI, error) {
	// Queue URL is like "https://sqs.us-west-2.amazonaws.com/415773754570/panther-alert-queue"
	region := strings.Split(queueURL, ".")[1]
	sqsClient, ok := client.sqsClients[region]
	if !ok {
		sqsClient = sqs.New(client.session, aws.NewConfig().WithRegion(region))
		client.sqsClients[region] = sqsClient
	}
	return sqsClient, nil
}
